posthog/temporal/batch_exports/transformer.py

```diff
@@ -55,243 +51,345 @@ def get_stream_transformer(
+                # In this case, we fallback to the slower but more permissive stdlib
+                # json.
+                logger.exception("Orjson detected a deeply nested dict: %s", d)
+                dumped = json.dumps(d, default=str).encode("utf-8") + b"\n"
+        else:
+            # Orjson is very strict about invalid unicode. This slow path protects us
+            # against things we've observed in practice, like single surrogate codes, e.g.
+            # "\ud83d"
+            logger.exception("Failed to encode with orjson: %s", d)
+            cleaned_content = replace_broken_unicode(d)
+            dumped = orjson.dumps(cleaned_content, default=str) + b"\n"
+
+    return dumped
+
+
+def replace_broken_unicode(obj):
+    if isinstance(obj, str):
+        return obj.encode("utf-8", "replace").decode("utf-8")
+    elif isinstance(obj, list):
+        return [replace_broken_unicode(item) for item in obj]
+    elif isinstance(obj, dict):
+        return {replace_broken_unicode(key): replace_broken_unicode(value) for key, value in obj.items()}
+    else:
+        return obj
+
+
+def json_dumps_bytes(d) -> bytes:
+    try:
+        return orjson.dumps(d, default=str)
+    except orjson.JSONEncodeError:
+        # orjson is very strict about invalid unicode. This slow path protects us against
+        # things we've observed in practice, like single surrogate codes, e.g. "\ud83d"
+        logger.exception("Failed to encode with orjson: %s", d)
+        cleaned_d = replace_broken_unicode(d)
+        return orjson.dumps(cleaned_d, default=str)
 
-                    try:
-                        await producer_task
-                    except asyncio.CancelledError:
-                        pass
 
+class ParquetStreamTransformer:
+    """A transformer to convert record batches into Parquet."""
 
-class ParquetStreamTransformer(StreamTransformer):
     def __init__(
         self,
         schema: pa.Schema,
         compression: str | None = None,
         compression_level: int | None = None,
         include_inserted_at: bool = False,
     ):
-        super().__init__(compression=compression, include_inserted_at=include_inserted_at)
+        self.include_inserted_at = include_inserted_at
         self.schema = schema
+        self.compression = compression
         self.compression_level = compression_level
 
         # For Parquet, we need to handle schema and batching
         self._parquet_writer: pq.ParquetWriter | None = None
-        self._parquet_buffer = BytesIO()
+        self._parquet_buffer = io.BytesIO()
+
+    async def iter(
+        self, record_batches: collections.abc.AsyncIterable[pa.RecordBatch], max_file_size_bytes: int = 0
+    ) -> collections.abc.AsyncGenerator[Chunk]:
+        """Iterate over record batches transforming them into chunks."""
+        current_file_size = 0
+
+        async for record_batch in record_batches:
+            chunk = await asyncio.to_thread(self.write_record_batch, record_batch)
```

Comment:
what's the thought behind using multiple threads? to prevent blocking the event loop in the main thread?

Reply:
Makes sense. Probably worth adding a comment just to mention why we're doing this